vllm/v1/worker/gpu_model_runner.py

```diff
@@ -2575,75 +2546,73 @@ def freeze_gc():
 
-    def _initialize_single_attn_backend(
-        self, kv_cache_spec: KVCacheSpec, layer_names: list[str]
-    ) -> tuple[AttentionBackend, AttentionMetadataBuilder]:
-        if isinstance(kv_cache_spec, AttentionSpec):
-            attn_backend_i = get_attn_backend(
-                kv_cache_spec.head_size,
-                self.dtype,
-                kv_cache_spec.dtype,
-                kv_cache_spec.block_size,
-                self.model_config.is_attention_free,
-                use_mla=kv_cache_spec.use_mla,
-            )
-            if attn_backend_i is None:
-                error_msg = (f"Error with get_attn_backend: "
-                             f"{kv_cache_spec.head_size=}, "
-                             f"{self.dtype=}, {kv_cache_spec.dtype=}, "
-                             f"{kv_cache_spec.block_size=}, "
-                             f"{self.model_config.is_attention_free=}, "
-                             f"{kv_cache_spec.use_mla=}")
-                logger.error(error_msg)
-                raise NotImplementedError(
-                    "Non-Attention backend is not supported by V1 "
-                    "GPUModelRunner.")
-        elif isinstance(kv_cache_spec, MambaSpec):
-            attn_backend_i = get_mamba_attn_backend(kv_cache_spec.mamba_type)
-        else:
-            raise ValueError(
-                f"Unknown KV cache spec type: {type(kv_cache_spec)}")
-
-        attn_metadata_builder_i = attn_backend_i.get_builder_cls()(
-            kv_cache_spec,
-            layer_names,
-            self.vllm_config,
-            self.device,
-        )
-
-        if (self.full_cuda_graph
-                and not attn_metadata_builder_i.full_cudagraph_supported):
-            raise ValueError(
-                f"Full CUDAGraph not supported for "
-                f"{attn_backend_i.__name__}. Turn off CompilationConfig."
-                f"full_cuda_graph or use a different attention backend.")
-        return attn_backend_i, attn_metadata_builder_i
-
     def initialize_attn_backend(self, kv_cache_config: KVCacheConfig) -> None:
         """
         Initialize the attention backends and attention metadata builders.
         """
-        assert len(self.attn_backends) == 0 and len(
-            self.attn_metadata_builders
-        ) == 0, "Attention backends are already initialized"
-        for i, kv_cache_group_spec in enumerate(
-                kv_cache_config.kv_cache_groups):
+        assert len(self.attn_groups) == 0, \
+            "Attention backends are already initialized"
+        attn_layers = get_layers_from_vllm_config(self.vllm_config, Attention)
+
+        def get_attn_backends_for_layers(
+                layer_names: list[str]
+        ) -> dict[type[AttentionBackend], list[str]]:
+            attn_backends = defaultdict(list)
+            for layer_name in layer_names:
+                attn_backends[attn_layers[layer_name].get_attn_backend(
+                )].append(layer_name)
```

Comment:
Because the chunked local attention backends are dynamically created unlike in `get_attn_backend(...)` which is cached, these will all be different objects (i.e. `hash(...)` will be different`). In case of Llama4 scout, you will see that `len(self.attn_groups[0])` is 37 instead of 2 (global and local) that we expect.

I thin we could dedupe here by `.__name__`, which would give us e.g. `[ChunkedLocalAttentionFlashAttentionBackend, FlashAttentionBackend]`. WDYT?

`__class__` doesn't work because they would all be `abc.ABCMeta`, similarly for `type()`. Using `functools.cache` decorator on `subclass_attention_metadata_builder` and `subclass_attention_backend` wouldn't work either since the types are dynamically created.

```suggestion
                attn_backends[attn_layers[layer_name].get_attn_backend(
                ).__name__].append(layer_name)
```

Reply:
oof good catch! updated it to use the name (and made sure the name is unique across block and attention_chunk sizes; not any model uses different attention_chunk sizes currently)